File: CodingSamples/Java/Inheritance/AbstractClassTest/src/Banker.java
package banking;

//factory class
public class Banker {

	private static long nid;

	//static initializer execute once after the class is loaded into the memory
	//commonly used to initialize static fields of the class
	static{
		nid = System.currentTimeMillis() % 1000000;
	}

	public static Account openCurrentAccount() {
		CurrentAccount acc = new CurrentAccount();
		acc.id = ++nid + 100000000;
		return acc;
	}

	public static Account openSavingsAccount() {
		SavingsAccount acc = new SavingsAccount();
		acc.id = ++nid + 200000000;
		return acc;
	}

	//a class that only contains static members does not require new objects
	//or subclasses and as such this class should define a private constructor 
	private Banker(){}
}

File: CodingSamples/DotNet/Inheritance/InterfaceTest1/Banking/Banker.cs
namespace Banking;

//a class defined with static modifier can only contain 
//static members, it does not define a non-static constructor 
//and as such it cannot be used for activation or extension
public static class Banker
{
    private static long nid;

    //a static constructor is called when a class is used 
    //for the first time within the CLR
    static Banker()
    {
        nid = DateTime.Now.Ticks % 1000000;
    }

    public static Account OpenCurrentAccount()
    {
        var acc = new CurrentAccount(); //implicitly typed local (var = CurrentAccount)
        acc.Id = ++nid + 100000000;
        return acc;
    }

    public static Account OpenSavingsAccount()
    {
        SavingsAccount acc = new(); //new SavingsAccount()
        acc.Id = ++nid + 200000000;
        return acc;
    }

    //An 'extension method' is a member of a 'static class' whose first
    //parameter is modified with 'this' keyword. Such a method can be
    //called as an 'instance method' of its 'first parameter type' in a scope
    //within which its class can be used without the namespace qualifier.
    public static void FreezeAccount(this Account acc)
    {
        acc.Id = -acc.Id;
    }
}
